xla/pjrt/pjrt_stream_executor_client.cc

```diff
@@ -3481,17 +3481,23 @@ PjRtStreamExecutorClient::GetExecutableExtras(CompileOptions* options) {
   if (device_assignment != nullptr) {
     addressable_device_logical_ids.reserve(num_replicas * num_partitions);
     addressable_devices.reserve(num_replicas * num_partitions);
+    absl::flat_hash_set<int> all_process_indices;
+    std::optional<int> this_process_index;
     for (int replica = 0; replica < num_replicas; ++replica) {
       for (int partition = 0; partition < num_partitions; ++partition) {
         int64_t device_id = (*device_assignment)(replica, partition);
         PjRtGlobalDeviceId global_device_id(device_id);
 
         TF_ASSIGN_OR_RETURN(PjRtDevice * device,
                             LookupDevice(global_device_id));
+        all_process_indices.insert(device->process_index());
         if (device->process_index() != process_index()) {
           VLOG(3) << "Non-local device: " << device_id;
           continue;
         }
+        if (!this_process_index.has_value()) {
+          this_process_index = all_process_indices.size() - 1;
```

Comment:
Not sure I understood this line correctly.
Doesn't this mean that all processes will assign this_process_index by the index of the first device?

Reply:
Each process iterates over all participating devices, looks at the process index corresponding to the device and compares it to its own process id. Using the hash set it counts how many other processes it has seen before it finds own one. In other words, this assigns process index for compilation based on first occurrence of this process in the {device: process} map.

Reply:
Ah sorry, I didn't notice `if (device->process_index() != process_index()) {` line above, make sense.

Isn't this equal to just initializing `build_options.set_process_index(process_index())` ?

Reply:
No - for sharded autotuning to work we want to always re-enumerate participating processes for compilation to 0...count-1 disregarding their original indices because an arbitrary subset of them can be participating in a compilation.